File: js/public.js
/* =============================================================================
   PUBLIC.JS — Глобальное общество Fielsdown
   Подключается ко всем страницам. Добавляет блок "Общество" внизу.
   Показывает: всех пользователей, все доски — в одном месте.
   ============================================================================= */

(function () {
  // Не запускаем на страницах регистрации/входа
  const currentPath = window.location.pathname;
  if (currentPath.includes('register.html') || currentPath.includes('login.html')) {
    return;
  }

  // Ждём полной загрузки DOM
  if (document.readyState === 'loading') {
    document.addEventListener('DOMContentLoaded', initPublicSection);
  } else {
    initPublicSection();
  }

  function initPublicSection() {
    // Создаём контейнер для общества
    const publicSection = document.createElement('div');
    publicSection.id = 'fielsdown-public-section';
    publicSection.style.cssText = `
      margin-top: 60px;
      padding: 30px 0;
      border-top: 1px solid #e0e0e0;
      background: #fafafa;
    `;

    publicSection.innerHTML = `
      <div class="container" style="max-width: 1000px; margin: 0 auto; padding: 0 20px;">
        <h2 style="font-size: 24px; color: #0077ff; margin-bottom: 24px; text-align: center;">
          🌍 Общество Fielsdown
        </h2>
        <div style="display: flex; gap: 30px; flex-wrap: wrap;">
          <!-- Пользователи -->
          <div style="flex: 1; min-width: 300px;">
            <h3 style="font-size: 18px; margin-bottom: 16px; color: #121212;">Пользователи</h3>
            <div id="public-users-list" style="display: grid; gap: 12px;"></div>
          </div>
          <!-- Доски -->
          <div style="flex: 1; min-width: 300px;">
            <h3 style="font-size: 18px; margin-bottom: 16px; color: #121212;">Доски сообщества</h3>
            <div id="public-boards-list" style="display: grid; gap: 12px;"></div>
          </div>
        </div>
      </div>
    `;

    // Вставляем перед </body>
    document.body.appendChild(publicSection);

    // Загружаем данные
    loadPublicData();
  }

  // === Загрузка данных ===
  function loadPublicData() {
    loadUsers();
    loadBoards();
  }

  function safeGet(key) {
    try {
      const raw = localStorage.getItem(key);
      return raw ? JSON.parse(raw) : null;
    } catch (e) {
      return null;
    }
  }

  const DEFAULT_AVATAR = 'https://static.cdninstagram.com/rsrc.php/v3/yo/r/qhYsMwhQJy-.png';

  function formatDate(dateStr) {
    if (!dateStr) return '';
    const date = new Date(dateStr);
    const diffHours = (Date.now() - date) / (1000 * 60 * 60);
    if (diffHours < 24) return 'сегодня';
    if (diffHours < 48) return 'вчера';
    return date.toLocaleDateString('ru-RU', { day: '2-digit', month: 'short' });
  }

  // === Пользователи ===
  function loadUsers() {
    const users = safeGet('fielsdown_users_v1') || {};
    const userList = Object.values(users);
    const container = document.getElementById('public-users-list');

    if (userList.length === 0) {
      container.innerHTML = '<p style="color:#666;font-style:italic;">Пока нет пользователей.</p>';
      return;
    }

    // Сортировка: новые выше
    userList.sort((a, b) => new Date(b.createdAt) - new Date(a.createdAt));

    const html = userList.map(user => `
      <div style="
        display: flex;
        align-items: center;
        padding: 10px;
        background: white;
        border: 1px solid #e0e0e0;
        border-radius: 8px;
      ">
        <a href="/profile.html?user=${encodeURIComponent(user.username)}" style="text-decoration:none;">
          <img 
            src="${user.avatar || DEFAULT_AVATAR}" 
            style="width:32px;height:32px;border-radius:50%;object-fit:cover;"
            onerror="this.src='${DEFAULT_AVATAR}'"
          >
        </a>
        <div style="margin-left:12px;">
          <div style="font-weight:600;color:#0077ff;font-size:15px;">
            b/${user.username}
          </div>
          <div style="font-size:12px;color:#666;">${formatDate(user.createdAt)}</div>
        </div>
      </div>
    `).join('');

    container.innerHTML = html;
  }

  // === Доски ===
  function loadBoards() {
    const boards = safeGet('fielsdown_boards_v1') || [];
    const container = document.getElementById('public-boards-list');

    if (boards.length === 0) {
      container.innerHTML = '<p style="color:#666;font-style:italic;">Пока нет досок.</p>';
      return;
    }

    boards.sort((a, b) => new Date(b.createdAt) - new Date(a.createdAt));

    const html = boards.map(board => `
      <div style="
        padding: 12px;
        background: white;
        border: 1px solid #e0e0e0;
        border-radius: 8px;
      ">
        <a 
          href="/board.html?b=${encodeURIComponent(board.name)}" 
          style="text-decoration:none;font-weight:bold;color:#0077ff;font-size:16px;"
        >
          b/${board.name}
        </a>
        <p style="margin:8px 0 6px 0;color:#121212;font-size:14px;">${board.description || 'Без описания.'}</p>
        <div style="font-size:12px;color:#666;">
          by b/${board.creator || 'anonymous'} • ${formatDate(board.createdAt)}
        </div>
      </div>
    `).join('');

    container.innerHTML = html;
  }
})();

File: public.go
package main

import (
	"encoding/json"
	"html/template"
	"log"
	"net/http"
	"sync"
	"time"

	"github.com/gorilla/mux"
)

// Board — структура доски
type Board struct {
	ID          int    `json:"id"`
	Name        string `json:"name"`
	Description string `json:"description"`
	Creator     string `json:"creator"`
	CreatedAt   string `json:"created_at"`
}

// In-memory storage
var (
	boards = []Board{}
	mutex  = sync.RWMutex{}
	nextID = 1
)

// createBoard — создаёт доску
func createBoard(w http.ResponseWriter, r *http.Request) {
	var input struct {
		Name        string `json:"name"`
		Description string `json:"description"`
		Creator     string `json:"creator"`
	}

	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		http.Error(w, "Неверный JSON", http.StatusBadRequest)
		return
	}

	if input.Name == "" {
		http.Error(w, "Укажите название", http.StatusBadRequest)
		return
	}

	mutex.RLock()
	for _, b := range boards {
		if b.Name == input.Name {
			mutex.RUnlock()
			http.Error(w, "Доска уже существует", http.StatusConflict)
			return
		}
	}
	mutex.RUnlock()

	mutex.Lock()
	board := Board{
		ID:          nextID,
		Name:        input.Name,
		Description: input.Description,
		Creator:     input.Creator,
		CreatedAt:   time.Now().Format("02.01.2006"),
	}
	boards = append(boards, board)
	nextID++
	mutex.Unlock()

	// Перенаправляем на board.html
	http.Redirect(w, r, "/board.html?b="+input.Name, http.StatusSeeOther)
}

// getIndex — отдаёт index.html с досками
func getIndex(w http.ResponseWriter, r *http.Request) {
	mutex.RLock()
	boardData := make([]map[string]interface{}, len(boards))
	for i, b := range boards {
		boardData[i] = map[string]interface{}{
			"Name":        b.Name,
			"Description": b.Description,
			"Creator":     b.Creator,
			"CreatedAt":   b.CreatedAt,
		}
	}
	mutex.RUnlock()

	tmpl := `
<!DOCTYPE html>
<html lang="ru">
<head>
  <meta charset="UTF-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Fielsdown — Fandom Community</title>
  <link rel="stylesheet" href="/css/globals.css" />
  <link rel="icon" href="/public/favicon.ico" />
  <style>
    .boards-grid { display: grid; grid-template-columns: repeat(auto-fill, minmax(280px, 1fr)); gap: 20px; margin-bottom: 40px; }
    .board-card { background: #fafafa; border: 1px solid #e0e0e0; border-radius: 12px; padding: 16px; }
    .board-link { display: block; font-size: 18px; font-weight: 700; margin-bottom: 6px; color: #121212; text-decoration: none; }
    .board-prefix { color: #0077ff; font-weight: 800; }
    .board-desc { font-size: 14px; color: #666; line-height: 1.5; }
    .board-create { display: flex; flex-direction: column; align-items: center; justify-content: center; color: #999; font-weight: 600; font-size: 18px; text-decoration: none; background: #fafafa; border: 1px solid #e0e0e0; border-radius: 12px; padding: 16px; }
    .plus-sign { font-size: 32px; font-weight: bold; margin-bottom: 8px; color: #0077ff; }
  </style>
</head>
<body>
  <header class="topbar">
    <div class="container">
      <h1 class="logo">Fielsdown</h1>
      <nav class="tabs"><a href="/" class="tab active">Посты</a></nav>
      <div class="auth-section" id="auth-section">
        <!-- Сессия будет обновляться через JS -->
      </div>
    </div>
  </header>

  <main class="container">
    <div class="boards-list">
      <h2>Доски сообщества</h2>
      <div class="boards-grid" id="boards-container">
        {{range .Boards}}
        <div class="board-card">
          <a href="/board.html?b={{.Name}}" class="board-link">
            <span class="board-prefix">b/</span>{{.Name}}
          </a>
          <p class="board-desc">{{.Description}}</p>
        </div>
        {{end}}
        <a href="/thread.html" class="board-card board-create">
          <span class="plus-sign">+</span>
          <span>Создать доску</span>
        </a>
      </div>
    </div>
  </main>

  <footer class="footer">
    <div class="container">
      <p>Fielsdown © 2025 — Fandom Community без цензуры и шаблонов.</p>
    </div>
  </footer>

  <script>
    // Сохраняем старую логику сессии из auth.js
    function getSession() {
      try {
        const raw = localStorage.getItem('fielsdown_session_v1');
        return raw ? JSON.parse(raw) : null;
      } catch (e) {
        return null;
      }
    }

    function updateAuthUI() {
      const authSection = document.getElementById('auth-section');
      if (!authSection) return;

      const session = getSession();
      const DEFAULT_AVATAR = 'https://static.cdninstagram.com/rsrc.php/v3/yo/r/qhYsMwhQJy-.png';

      if (session && session.isLoggedIn) {
        authSection.innerHTML = 
          '<div class="user-menu" onclick="window.location.href=\\'/profile.html\\'" style="display:flex;align-items:center;gap:12px;cursor:pointer;">' +
          '<img src="' + (session.avatar || DEFAULT_AVATAR) + '" style="width:36px;height:36px;border-radius:50%;object-fit:cover;border:2px solid #e0e0e0;background:#f5f5f5;" onerror="this.src=\\'' + DEFAULT_AVATAR + '\\'">' +
          '<span class="user-nick" style="font-weight:600;color:#121212;font-size:15px;">b/' + session.username + '</span>' +
          '</div>';
      } else {
        authSection.innerHTML = 
          '<a href="/register.html" class="btn btn-secondary">Регистрация</a>' +
          '<a href="/login.html" class="btn btn-secondary">Вход</a>';
      }
    }

    document.addEventListener('DOMContentLoaded', updateAuthUI);
  </script>
</body>
</html>
`

	data := struct {
		Boards []map[string]interface{}
	}{
		Boards: boardData,
	}

	t, _ := template.New("index").Parse(tmpl)
	t.Execute(w, data)
}

// corsMiddleware
func corsMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Access-Control-Allow-Origin", "*")
		w.Header().Set("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
		w.Header().Set("Access-Control-Allow-Headers", "Content-Type")
		if r.Method == "OPTIONS" {
			w.WriteHeader(http.StatusNoContent)
			return
		}
		next.ServeHTTP(w, r)
	})
}

func main() {
	r := mux.NewRouter()

	// API
	r.HandleFunc("/api/boards", createBoard).Methods("POST")

	// Страницы
	r.HandleFunc("/", getIndex).Methods("GET")
	r.PathPrefix("/").Handler(http.FileServer(http.Dir("./")))

	handler := corsMiddleware(r)

	log.Println("Сервер запущен на :8080")
	log.Fatal(http.ListenAndServe(":8080", handler))
}
